src/main/java/uk/gov/hmcts/ccd/domain/service/common/CompoundAccessControlService.java

```diff
@@ -0,0 +1,269 @@
+                    break;
+            }
+        } else if (caseField.isComplexFieldType() && data != null) {
+            createDenied = isCreateDeniedForAnyChildNode(caseField, userRoles, data);
+        }
+        return createDenied;
+    }
+
+    private boolean isCreateDeniedForAnyChildNode(final CaseField caseField, final Set<String> userRoles, final JsonNode node) {
+        boolean createDeniedForAnyChild = false;
+        if (caseField.isCollectionFieldType()) {
+            for (CaseField field : caseField.getFieldType().getCollectionFieldType().getComplexFields()) {
+                if (field.isCompound() && node.get(VALUE) != null && node.get(VALUE).get(field.getId()) != null
+                    && isCreateDenied(node.get(VALUE).get(field.getId()), field, userRoles)) {
+                    LOG.info("Child {} of {} has new data and no create access", field.getId(), caseField.getId());
+                    createDeniedForAnyChild = true;
+                    break;
+                }
+            }
+        } else {
+            for (CaseField field : caseField.getFieldType().getComplexFields()) {
+                if (field.isCompound() && isCreateDenied(node.get(field.getId()), field, userRoles)) {
+                    LOG.info("Child {} has new data and no create access", field.getId());
+                    createDeniedForAnyChild = true;
+                    break;
+                }
+            }
+        }
+        return createDeniedForAnyChild;
+    }
+
+    private boolean itemDeletedAndHasDeleteAccess(JsonNode existingData, JsonNode newData, CaseField caseField, Set<String> userRoles) {
+        boolean containsDeletedItem = caseField.isCollectionFieldType() && StreamSupport
+            .stream(spliteratorUnknownSize(existingData.get(caseField.getId()).elements(), Spliterator.ORDERED), false)
+            .anyMatch(oldItem -> itemMissing(oldItem, newData.get(caseField.getId())));
+
+        return (!containsDeletedItem || hasAccessControlList(userRoles, CAN_DELETE, caseField.getAccessControlLists()))
+            && !isDeleteDeniedForChildren(existingData.get(caseField.getId()), newData.get(caseField.getId()), caseField, userRoles);
+    }
+
+    private boolean itemMissing(JsonNode oldItem, JsonNode newValue) {
+        boolean itemExists = StreamSupport.stream(spliteratorUnknownSize(newValue.elements(), Spliterator.ORDERED), false)
+            .anyMatch(newItem -> !isNullId(newItem) && newItem.get("id").equals(oldItem.get("id")));
+        return !itemExists;
+    }
+
+    private boolean isDeleteDeniedForChildren(final JsonNode existingData, final JsonNode newData, final CaseField caseField, final Set<String> userRoles) {
+        boolean deleteDenied = false;
+        if (caseField.isCollectionFieldType() && existingData.size() > 0) {
+            for (JsonNode existingNode : existingData) {
+                deleteDenied = isCurrentNodeOrAnyChildNodeDeletedWithoutAccess(newData, caseField, userRoles, existingNode);
```

Comment:
same as above if it make sense / works 😄

Reply:
👍